### PR TITLE
Add default property value converters for all value types

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

It is possible to create property editors without backing property value converters. Usually this is done using manifests. These property editors still declare a value type, which should be one of the predefined ones in [`ValueTypes`](https://github.com/umbraco/Umbraco-CMS/blob/v13/dev/src/Umbraco.Core/PropertyEditors/ValueTypes.cs).

However, the resulting property values for these editor types are not always typed according to their declared value types. Rather, their types tie in part into their storage column in the database, and in part into the de/serialization that goes on within NuCache:

- Value type DECIMAL becomes a `string` instead of a `decimal` (even through it is backed by a decimal column in the database, it still becomes a `string` due to a [known issue with MessagePack](https://github.com/MessagePack-CSharp/MessagePack-CSharp/issues/651)
- Value type TIME becomes a `string` instead of a `TimeSpan`.
- Value type BIGINT becomes a `string` instead of a `long`.
- Value type XML becomes a `string` instead of an `XDocument`.

image

As for ModelsBuilder, all these editor types produce `object` properties on the generated model classes - except for value type JSON, which produces `JToken` properties due to the default [JsonValueConverter](https://github.com/umbraco/Umbraco-CMS/blob/v13/dev/src/Umbraco.Infrastructure/PropertyEditors/ValueConverters/JsonValueConverter.cs#L49):

image

This PR introduces default value converters to handle the conversion of all value types (except JSON, as it's already handled). Thus, value types are translated into corresponding types:

image

...also in the generated models:

image

### Testing this PR

It's quite cumbersome to test this PR in detail.

First and foremost you need custom property editors (without property value converters) defined for each value type, to test the resulting value outputs and generated models.

Then, due to how this PR amends the "shadowing" logic of property value converters, you need to create a doctype which contains every single built-in property and ensure that they render their values correctly. In other words, none of these properties can be left empty - not even the Label properties, which in turn means you'll need a notification handler to add values to the Label properties (since they're not editable).

Long story short - it might be best to peer test this with @kjac who has all of this setup and ready to go 😆 

### Breaking change?

This is in principle both a binary and a behavioural a breaking change. Property value types change and so do the generated models.

### CODE REVIEW

**Comment by bjarnef** [comment]
Oh, so it seems this may fix this old reported issue/feature request @kjac ? 😅
https://github.com/umbraco/Umbraco-CMS/issues/4419

Off topic, but related to generated types I would like if it was possible to define nullable property, especially with numbers (`int` or `decimal` where 0 may be a valid value). Regarding `DateTime` often default value `DateTime.MinValue` isn't something editors would select, so it may be a workaround to compare with in most cases:
https://github.com/umbraco/Umbraco-CMS/discussions/13816

**Comment by kjac** [comment]
Great notes, @bjarnef ❤️ 

Changing the generated nullability based on type detection and possibly mandatory configuration is not in scope for this change. It's an interesting proposal, though.

While this *is* a breaking change for some, we still try to keep the breakage at a minimum in this change.

**Comment by Zeegaan** [comment]
Looks good, tests good 🚀

**Comment by jaandrews** [comment]
Just a note for anyone encountering issues due to this change, the problem is that properties with custom value converters will now have two value converters that match them if you filter them based on the property type, the default converter and the custom one. A pattern I often use to reference a particular converter is
```csharp
    var converter = _propertyValueConverters.Value.FirstOrDefault(x => x.IsConverter(property.PropertyType));
```
The default converter is now the first result that this retrieves, so it won't return the expected converter. A simple fix for this is to get the last match instead.
```csharp
    var converter = _propertyValueConverters.Value.LastOrDefault(x => x.IsConverter(property.PropertyType));
```
